File: tools/read.py
from pathlib import Path

from .glob import suggest_files
from .textfile import read_text, split_lines


BINARY_EXTS = {
    ".ico", ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".webp", ".svgz",
    ".exe", ".dll", ".so", ".dylib", ".pdb", ".bin", ".obj", ".lib",
    ".zip", ".rar", ".7z", ".gz", ".tar", ".xz", ".jar", ".nupkg",
    ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
    ".mp3", ".mp4", ".avi", ".mkv", ".wav", ".mov", ".ttf", ".otf",
    ".woff", ".woff2", ".pyc", ".pyd", ".class", ".wasm", ".db", ".sqlite",
}

MAX_READ_BYTES = 400_000
MAX_LINE_CHARS = 2000


def _looks_binary(filepath: Path) -> bool:
    """A NUL byte in the first 8 KB is the classic binary heuristic."""
    try:
        with open(filepath, "rb") as f:
            return b"\x00" in f.read(8192)
    except OSError:
        return False


def read_file(path: str, offset: int = 1, limit: int = 2000) -> str:
    """Read a file from the filesystem and return its contents.

    Args:
        path: Path to the file (absolute or relative to workspace)
        offset: Line number to start reading from (1-indexed, default 1)
        limit: Maximum number of lines to read (default 2000, max 5000)
    """
    try:
        filepath = Path(path).resolve()
        if not filepath.exists():
            msg = f"Error: File not found: {path}"
            # Suggestions are scoped to the file's intended parent. Falling
            # back to cwd used to trigger a full recursive scan of whatever
            # directory the process happened to be in.
            parent = filepath.parent
            candidates = suggest_files(path, parent) if parent.is_dir() else []
            if candidates:
                msg += " Did you mean:\n" + "\n".join(f"  {c}" for c in candidates)
            return msg
        if not filepath.is_file():
            return f"Error: Not a file: {path}"

        # Binary content read as text becomes replacement-char noise that
        # poisons the context and sends small models into a repetition loop.
        if filepath.suffix.lower() in BINARY_EXTS:
            size = filepath.stat().st_size
            return (
                f"Error: '{filepath.name}' is a binary file ({_fmt_size(size)}), not text. "
                "Do not read it. Its name and size are all the information you need."
            )
        if _looks_binary(filepath):
            size = filepath.stat().st_size
            return (
                f"Error: '{filepath.name}' appears to be binary ({_fmt_size(size)}), not text. "
                "Do not read it."
            )

        size = filepath.stat().st_size
        if size > MAX_READ_BYTES:
            return (
                f"Error: '{filepath.name}' is too large ({_fmt_size(size)}). "
                "Use search_files to find the relevant part, then read with offset/limit."
            )

        offset = _as_int(offset, 1)
        limit = _as_int(limit, 2000)

        # Shared splitting keeps these line numbers identical to the ones
        # edit_file addresses. Reading in text mode here (and raw there) used
        # to make the two disagree on any file with mixed line endings, so an
        # edit landed on the wrong line.
        content, _ = read_text(filepath)
        all_lines, _ = split_lines(content)
        total = len(all_lines)

        cap = max(1, min(limit, 5000))
        start = max(0, offset - 1)
        end = start + cap

        if start >= total:
            return f"--- {filepath} (offset {offset} past end, {total} lines total) ---\n"

        lines = []
        for n in range(start, min(end, total)):
            line = all_lines[n]
            if len(line) > MAX_LINE_CHARS:
                line = line[:MAX_LINE_CHARS] + " ... [line truncated]"
            lines.append(f"{n + 1}: {line}\n")

        info = f"--- {filepath} (lines {start + 1}-{min(end, total)} of {total}) ---\n"
        if end < total:
            info += f"... (showing {len(lines)} of {total} lines, use offset={end + 1} for more)\n"

        return info + "".join(lines)
    except Exception as e:
        return f"Error reading file: {e}"


def _as_int(value, default: int) -> int:
    """Small models send "1" or 1.0 where the schema says integer."""
    if isinstance(value, bool) or value is None:
        return default
    if isinstance(value, int):
        return value
    try:
        return int(float(str(value).strip()))
    except (TypeError, ValueError):
        return default


def _fmt_size(size: int) -> str:
    if size < 1024:
        return f"{size}B"
    if size < 1024 * 1024:
        return f"{size / 1024:.1f}KB"
    return f"{size / 1024 / 1024:.1f}MB"

File: internal/tools/read.go
package tools

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// BINARY_EXTS are extensions never read as text.
var BINARY_EXTS = map[string]bool{
	".ico": true, ".png": true, ".jpg": true, ".jpeg": true, ".gif": true,
	".bmp": true, ".webp": true, ".svgz": true, ".exe": true, ".dll": true,
	".so": true, ".dylib": true, ".pdb": true, ".bin": true, ".obj": true,
	".lib": true, ".zip": true, ".rar": true, ".7z": true, ".gz": true,
	".tar": true, ".xz": true, ".jar": true, ".nupkg": true, ".pdf": true,
	".doc": true, ".docx": true, ".xls": true, ".xlsx": true, ".ppt": true,
	".pptx": true, ".mp3": true, ".mp4": true, ".avi": true, ".mkv": true,
	".wav": true, ".mov": true, ".ttf": true, ".otf": true, ".woff": true,
	".woff2": true, ".pyc": true, ".pyd": true, ".class": true, ".wasm": true,
	".db": true, ".sqlite": true,
}

const (
	maxReadBytes = 400000
	maxLineChars = 2000
)

// ReadFile reads a file and returns its contents with line numbers.
func ReadFile(workspace, path string, offset, limit int) string {
	filepathAbs := resolveUnder(workspace, path)
	info, err := os.Stat(filepathAbs)
	if os.IsNotExist(err) {
		msg := fmt.Sprintf("Error: File not found: %s", path)
		candidates := SuggestFiles(workspace, path, 5)
		if len(candidates) > 0 {
			msg += " Did you mean:\n" + strings.Join(candidates, "\n  ")
		}
		return msg
	}
	if err != nil {
		return fmt.Sprintf("Error reading file: %v", err)
	}
	if !info.Mode().IsRegular() {
		return fmt.Sprintf("Error: Not a file: %s", path)
	}

	if BINARY_EXTS[strings.ToLower(filepath.Ext(filepathAbs))] {
		return fmt.Sprintf("Error: '%s' is a binary file (%s), not text. Do not read it. Its name and size are all the information you need.",
			filepath.Base(filepathAbs), fmtSize(info.Size()))
	}

	data, err := os.ReadFile(filepathAbs)
	if err != nil {
		return fmt.Sprintf("Error reading file: %v", err)
	}
	if bytes.IndexByte(data, 0) >= 0 {
		return fmt.Sprintf("Error: '%s' appears to be binary (%s), not text. Do not read it.",
			filepath.Base(filepathAbs), fmtSize(info.Size()))
	}
	if len(data) > maxReadBytes {
		return fmt.Sprintf("Error: '%s' is too large (%s). Use search_files to find the relevant part, then read with offset/limit.",
			filepath.Base(filepathAbs), fmtSize(info.Size()))
	}

	if offset < 1 {
		offset = 1
	}
	if limit < 1 {
		limit = 2000
	}
	if limit > 5000 {
		limit = 5000
	}

	lines := strings.Split(string(data), "\n")
	total := len(lines)
	if total > 0 && lines[total-1] == "" {
		total--
	}

	start := offset - 1
	end := start + limit
	if start > total {
		return fmt.Sprintf("--- %s (offset %d past end, %d lines total) ---\n", filepathAbs, offset, total)
	}
	if end > total {
		end = total
	}

	var b strings.Builder
	fmt.Fprintf(&b, "--- %s (lines %d-%d of %d) ---\n", filepathAbs, start+1, end, total)
	for i := start; i < end; i++ {
		line := lines[i]
		if len(line) > maxLineChars {
			line = line[:maxLineChars] + " ... [line truncated]"
		}
		fmt.Fprintf(&b, "%d: %s\n", i+1, line)
	}
	if end < total {
		fmt.Fprintf(&b, "... (showing %d of %d lines, use offset=%d for more)\n", end-start, total, end+1)
	}
	return strings.TrimRight(b.String(), "\n")
}
